arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/raise/Builders.kt

```diff
@@ -304,14 +292,11 @@ public class OptionRaise(private val raise: Raise<None>) : Raise<None> by raise
  * Implementation of [Raise] used by [ior].
  * You should never use this directly.
  */
-public class IorRaise<Error> @PublishedApi internal constructor(
-  @PublishedApi internal val combineError: (Error, Error) -> Error,
-  private val state: Atomic<Option<Error>>,
+public abstract class IorRaise<Error> @PublishedApi internal constructor(
   private val raise: Raise<Error>,
-) : Raise<Error> {
-
-  @RaiseDSL
-  override fun raise(r: Error): Nothing = raise.raise(combine(r))
+) : Raise<Error> by raise {
+  @PublishedApi
+  internal abstract fun addError(e: Error)
```

Comment:
I think this function should still be called `combine`, there's no need to break API compatibility.

Reply:
I didn't want the `Atomic` to be part of `IorRaise` because it's an implementation detail that it doesn't need, same with `combine`. `IorRaise` just needs a general place to add extra errors into whenever there's an `Ior.Both`. I agree though that ABI compatibility is important, so I'll update this so the ABI is identical.
Also, the issue with exposing a `combine` is that the previous implementation erroneously used it inside `raise`, which prevents `try-catch` from recovering from it. I'll add a test for that as well. There's no real reason why `IorRaise` ever needs to read the accumulated errors so far. The `ior` builder needs to have that info, of course, but only as an implementation detail.